src/runtime/ranger/ranger_resource_policy_manager.cpp

```diff
@@ -171,5 +215,382 @@ void ranger_resource_policy_manager::parse_policies_from_json(const rapidjson::V
+
+    update_cached_policies();
+    LOG_DEBUG("Update using resources policies succeed.");
+
+    if (dsn::ERR_OK == sync_policies_to_app_envs()) {
+        LOG_ERROR("Sync policies to app envs failed.");
+    } else {
+        LOG_DEBUG("Sync policies to app envs succeeded.");
+    }
+}
+
+void ranger_resource_policy_manager::dump_policies_to_remote_storage()
+{
+    dsn::blob value = json::json_forwarder<all_resource_policies>::encode(_all_resource_policies);
+    _meta_svc->get_remote_storage()->set_data(
+        _ranger_policy_meta_root, value, LPC_USE_RANGER_ACCESS_CONTROL, [this](dsn::error_code e) {
+            if (e == dsn::ERR_OK) {
+                LOG_DEBUG("Dump Ranger policies to remote storage succeed.");
+                return;
+            }
+            if (e == dsn::ERR_TIMEOUT) {
+                LOG_ERROR("Dump Ranger policies to remote storage timeout, retry later.");
+                dsn::tasking::enqueue(LPC_USE_RANGER_ACCESS_CONTROL,
+                                      &_tracker,
+                                      [this]() { dump_policies_to_remote_storage(); },
+                                      0,
+                                      kLoadRangerPolicyRetryDelayMs);
+                return;
+            }
+            LOG_ERROR("Dump Ranger policies to remote storage failed.");
+        });
+}
+
+void ranger_resource_policy_manager::update_cached_policies()
+{
+    {
+        utils::auto_write_lock l(_global_policies_lock);
+        _global_policies_cache.swap(_all_resource_policies[enum_to_string(resource_type::kGlobal)]);
+        // TODO(wanghao): provide a query method
+    }
+    {
+        utils::auto_write_lock l(_database_policies_lock);
+        _database_policies_cache.swap(
+            _all_resource_policies[enum_to_string(resource_type::kDatabase)]);
+        // TODO(wanghao): provide a query method
+    }
+}
+
+dsn::error_code ranger_resource_policy_manager::sync_policies_to_app_envs()
+{
+    const auto &table_policies =
+        _all_resource_policies.find(enum_to_string(resource_type::kDatabaseTable));
+    if (table_policies == _all_resource_policies.end()) {
+        LOG_INFO("DATABASE_TABLE level policy is empty, skip to sync app envs.");
+        return dsn::ERR_OK;
+    }
+
+    dsn::replication::configuration_list_apps_response list_resp;
+    dsn::replication::configuration_list_apps_request list_req;
+    list_req.status = dsn::app_status::AS_AVAILABLE;
+    _meta_svc->get_server_state()->list_apps(list_req, list_resp);
+    ERR_LOG_AND_RETURN_NOT_OK(list_resp.err, "list_apps failed.");
+    for (const auto &app : list_resp.infos) {
+        std::string database_name = get_database_name_from_app_name(app.app_name);
+        std::string table_name;
+        if (database_name.empty()) {
+            database_name = "*";
+            table_name = app.app_name;
+        } else {
+            table_name = app.app_name.substr(database_name.size());
+        }
```

Comment:
```suggestion
        std::string table_name = get_table_name_from_app_name(app.app_name);
        if (database_name.empty()) {
            database_name = "*";
        }
```

Reply:
If `get_database_name_from_app_name` and `get_table_name_from_app_name` are always apear in pair, consider to encapsulate a function like `std::pair<std::string, std::string> get_db_table_name_from_app_name()` instead ?

Reply:
Keep it separate, just get the `database_name` at the database level ACL.